@ybogomolov: homework2 implementation

YBogomolov opened the pull request:
All present tests are passing. I am a bit uncertain about `zipListOrd/Num/Frac`'s `compare` method implementation, though. Especially part where a single-valued `Repeat` is compared against a `Finite`.

I would like @Odomontois to take a look at it.

YBogomolov committed:
Make Eq tests pass

YBogomolov committed:
Make all tests pass

YBogomolov committed:
Remove Num/Frac ops in favour of usage direct method calls on implicits. This helps resolving ambiguity of Ord/Num/Frac implicits when using <=> syntax.

catostrophe commented on src/main/scala/fpspeedrun/Monoid.scala:
```diff
@@ -4,39 +4,52 @@ import simulacrum.typeclass
+  implicit def endoMonoid[A]: Monoid[Endo[A]] = new Monoid[Endo[A]] {
+    override def empty: Endo[A] = Endo(identity)
+    override def combine(x: Endo[A], y: Endo[A]): Endo[A] =
+      Endo(a => y.run(x.run(a)))
```
see `andThen`

catostrophe commented on src/main/scala/fpspeedrun/syntax/package.scala:
```diff
@@ -47,14 +48,18 @@ object semigroup extends Semigroup.ToSemigroupOps {
-    def foldAll(implicit mon: Monoid[A]): A = ???
+  implicit class ListOps[A](val xs: List[A]) extends AnyVal {
+    def foldAll(implicit mon: Monoid[A]): A =
+      xs.foldRight(mon.empty)(mon.combine)
```
std `foldRight` on a list is actually `reverse` and `foldLeft`

catostrophe commented on src/main/scala/fpspeedrun/ZipList.scala:
```diff
@@ -9,33 +14,133 @@ trait ZipList[A] {
+
+  implicit def zipListOrd[A: Ord]: Ord[ZipList[A]] =
+    (x: ZipList[A], y: ZipList[A]) =>
+      x.value match {
```
patmatching on a tuple looks clearer/shorter here and below

YBogomolov committed:
Improvements after the review:
1. Direct function composition is replaced with `andThen` call.
2. `foldRight` is replaced with `foldLeft`.
3. Nested pattern matching is replaced with tuple pattern matching in `compare` methods.

YBogomolov committed:
Use syntax instead of direct `ops` import.
